fortuna/prob_model/predictive/base.py

```diff
@@ -49,6 +49,85 @@ def __init__(self, posterior: Posterior):
         self.likelihood = posterior.joint.likelihood
         self.posterior = posterior
 
+    ####
+    def ensemble_log_prob(
+        self,
+        data_loader: DataLoader,
+        n_posterior_samples: int = 30,
+        rng: Optional[PRNGKeyArray] = None,
+        distribute: bool = True,
+        **kwargs,
+    ) -> jnp.ndarray:
+        r"""
+        Compute the log-likelihood at each posterior sample, that is
+
+        .. math::
+            \log p(y|x, theta^{(i)}),
+
+        where:
+         - :math:`x` is an observed input variable;
+         - :math:`y` is an observed target variable;
+         - :math:`theta^{(i)}` is a sample from the posterior.
+
+        Parameters
+        ----------
+        data_loader : DataLoader
+            A data loader.
+        n_posterior_samples : int
+            Number of posterior samples to draw in order to compute the log -ikelihood.
+            that would be produced using the posterior distribution state.
+        rng : Optional[PRNGKeyArray]
+            A random number generator. If not passed, this will be taken from the attributes of this class.
+        distribute: bool
+            Whether to distribute computation over multiple devices, if available.
+
+        Returns
+        -------
+        jnp.ndarray
+            An array of log-likelihood values at each posterior sample for each data point.
+        """
+
+        if rng is None:
+            rng = self.rng.get()
+
+        log_probs = self._loop_fun_through_data_loader(
+            self._batched_ensemble_log_prob,
+            data_loader,
+            n_posterior_samples,
+            rng,
+            distribute,
+            **kwargs,
+        )
+
+        return log_probs.swapaxes(0, 1)
+
+    def _batched_ensemble_log_prob(
+        self,
+        batch: Batch,
+        n_posterior_samples: int = 30,
+        rng: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ) -> Union[jnp.ndarray, Tuple[jnp.ndarray, dict]]:
+        if rng is None:
+            rng = self.rng.get()
+        keys = random.split(rng, n_posterior_samples)
+
+        def _lik_log_batched_prob(key):
+            sample = self.posterior.sample(inputs=batch[0], rng=key)
+            return self.likelihood._batched_log_prob(
+                sample.params,
+                batch,
+                mutable=sample.mutable,
+                calib_params=sample.calib_params,
+                calib_mutable=sample.calib_mutable,
+                **kwargs,
+            )
+
+        return lax.map(_lik_log_batched_prob, keys).swapaxes(
```

Comment:
It looks like we swap axes both here and in `ensemble_log_prob`. Is this double swapping necessary?

Reply:
I think so - I tried removing it and it gave me concatenation errors. I believe it's due to the concatenation that occurs when running self._loop_fun_through_data_loader?